### ISSUE TITLE
NoSuchMethodError: method not found: 'toString' on null on web

### ISSUE THREAD

**Issue by rbellens** [dart-lang/sdk#46417]
The following code throws an error when executing on a web platform (tested on chrome and node)

```dart
var v = {...Map.unmodifiable({'x': null})};
```

This code runs fine on vm and also with dartdevc.

I'm using `Dart SDK version: 2.13.3 (stable) (Wed Jun 9 12:44:44 2021 +0200) on "macos_x64"`

**Comment by lrhn** [comment]
The error is:
```
Uncaught TypeError: Cannot read property 'toString' of nullError: TypeError: Cannot read property 'toString' of null
```
It happens during the spread operation.
```dart
void main() {
  var v = {...Map.unmodifiable({'x': null}.l).l}.l;
  v.l;
}
extension<T> on T {
  T get l {
    print("${this} : ${this.runtimeType} @ $T");
    return this;
  }
}
```
prints
```
{x: null} : JsLinkedHashMap<String, Null> @ Map<String, Null>
{x: null} : ConstantStringMap<dynamic, dynamic> @ Map<dynamic, dynamic>
Uncaught TypeError: Cannot read property 'toString' of nullError: TypeError: Cannot read property 'toString' of null
```

That suggests a stray non-null-check being applied to the map value during the spreading. I believe `x.toString` is being used as an "assert-not-`null`" check by dart2js compiled code.

**Comment by fishythefish** [comment]
From looking at the generated JS, I think it's a spurious `!` here: https://github.com/dart-lang/sdk/blob/35b94197eaab30b27d5d59e718b0fa67ba79d8dd/sdk/lib/_internal/js_runtime/lib/constant_map.dart#L77

**Comment by lrhn** [comment]
That looks very much like a bug and a likely cause of this error.

**Comment by fishythefish** [comment]
Fixed by c015e02180d0e4231eacc71817f6230182203bc2.
